bfvmm/src/vmcs/src/vmcs_intel_x64.cpp

```diff
@@ -454,37 +454,28 @@ vmcs_intel_x64::primary_processor_based_vm_execution_controls()
 void
 vmcs_intel_x64::secondary_processor_based_vm_execution_controls()
 {
-    auto controls = vm::read(VMCS_SECONDARY_PROCESSOR_BASED_VM_EXECUTION_CONTROLS);
-
-    // controls |= VM_EXEC_S_PROC_BASED_VIRTUALIZE_APIC_ACCESSES;
-    // controls |= VM_EXEC_S_PROC_BASED_ENABLE_EPT;
-    // controls |= VM_EXEC_S_PROC_BASED_DESCRIPTOR_TABLE_EXITING;
-    controls |= VM_EXEC_S_PROC_BASED_ENABLE_RDTSCP;
-    // controls |= VM_EXEC_S_PROC_BASED_VIRTUALIZE_X2APIC_MODE;
-    // controls |= VM_EXEC_S_PROC_BASED_ENABLE_VPID;
-    // controls |= VM_EXEC_S_PROC_BASED_WBINVD_EXITING;
-    // controls |= VM_EXEC_S_PROC_BASED_UNRESTRICTED_GUEST;
-    // controls |= VM_EXEC_S_PROC_BASED_APIC_REGISTER_VIRTUALIZATION;
-    // controls |= VM_EXEC_S_PROC_BASED_VIRTUAL_INTERRUPT_DELIVERY;
-    // controls |= VM_EXEC_S_PROC_BASED_PAUSE_LOOP_EXITING;
-    // controls |= VM_EXEC_S_PROC_BASED_RDRAND_EXITING;
-    controls |= VM_EXEC_S_PROC_BASED_ENABLE_INVPCID;
-    // controls |= VM_EXEC_S_PROC_BASED_ENABLE_VM_FUNCTIONS;
-    // controls |= VM_EXEC_S_PROC_BASED_VMCS_SHADOWING;
-    // controls |= VM_EXEC_S_PROC_BASED_RDSEED_EXITING;
-    // controls |= VM_EXEC_S_PROC_BASED_EPT_VIOLATION_VE;
-
-    // @connojd this needs to be re-enabled once these are fixed.
-    // These should all be enabled_if_allowed(), and then RDTSCP, INVPCID,
-    // and SXAVE should all be uncommented, with the rest commented.
-    // I commented out xsave for now so that the hypervisor will still
-    // work on VMWare for now.
-    //
-    // controls |= VM_EXEC_S_PROC_BASED_ENABLE_XSAVES_XRSTORS;
-
-    this->filter_unsupported(msrs::ia32_vmx_procbased_ctls2::addr, controls);
-
-    vm::write(VMCS_SECONDARY_PROCESSOR_BASED_VM_EXECUTION_CONTROLS, controls);
+    using namespace vmcs::secondary_processor_based_vm_execution_controls;
+
+    bool verbose = true;
+
+    // virtualize_apic_accesses::enable_if_allowed(verbose);
+    // enable_ept::enable_if_allowed(verbose);
+    // descriptor_table_exiting::enable_if_allowed(verbose);
+    enable_rdtscp::enable_if_allowed(verbose);
```

Comment:
Ok.... so this is where the issue is. 

There is a difference between allowed, and exists. The MSRs state if a control is allowed to be 0 or 1. You can actually have a control that exists, but is not allowed to be 0 or 1. A good example of that is exit on write to CR3. That has always existed, but was not allowed to be 0 until the TRUE based controls came out. 

So I like your naming scheme, as it handles both cases well. You can only enable / disable if it is both allowed and exists, and you can only query is it exists (allowed doesn't matter).

So +1